presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java

```diff
@@ -176,8 +184,19 @@ private void pollWorkers()
 
         // Add new nodes
         for (InternalNode node : aliveNodes) {
-            nodeStates.putIfAbsent(node.getNodeIdentifier(),
-                    new HttpRemoteNodeState(httpClient, uriBuilderFrom(node.getInternalUri()).appendPath("/v1/info/state").build()));
+            if (protocol.equals(CommunicationProtocol.THRIFT) && node.getThriftPort().isPresent()) {
+                RemoteNodeState remoteNodeState = nodeStates.get(node.getNodeIdentifier());
+                // TODO: hack; thrift announcement is slower than http ones
+                if (!(remoteNodeState instanceof ThriftRemoteNodeState)) {
```

Comment:
Is my understanding correct that we are trying to guard against the situation where a node has changed to use thrift, but the announcement has not made it so the internal state management has not updated the state yet? 

That being said, theoretically it is also possible if we have a thrift -> http change? Do we need to handle that? Or is my understanding totally off?

Reply:
Your understand is correct. In theory no as long as the protocol is in config; so we always stick to one protocol.

Reply:
> but the announcement has not made it so the internal state management has not updated the state yet?

Maybe I didn't understand correctly... why would this happen -- if the config is consistent? 

When I read the comment, "// TODO: hack; thrift announcement is slower than http ones", my understanding is worker will both announce http and thrift announcement?  -- it seems to defeat the purpose to have a more efficient RPC-based announcement?

Reply:
Let me rewrite the if/else to make it clearer.